### ISSUE TITLE
Support for long paths

### ISSUE THREAD

**Issue by justanotheranonymoususer** [rodrigocfd/winsafe#145]
Have you considered supporting longs paths beyond `MAX_PATH` for functions such as `GetSystemDirectory`, `GetModuleFileName` and many others?

You can see how Microsoft's WIL (C++) does it by probing for the file size:
https://github.com/microsoft/wil/blob/27d66a24b55a2f7baf91c06c196ba1a2d256c2db/include/wil/win32_helpers.h#L544-L557

**Comment by rodrigocfd** [comment]
Oh I've use this technique before in [`ListViewItem::text`](https://rodrigocfd.github.io/winsafe/winsafe/gui/spec/struct.ListViewItem.html#method.text).

I've implemented these functions, whose documentation exhibit the same behavior:

* GetSystemDirectory
* GetModuleFileName
* GetCurrentDirectory

If you need another one, feel free to open another issue.

**Comment by justanotheranonymoususer** [comment]
Thanks. GetModuleFileName is a bad API, and indeed you have no choice but try blindly. MS [use *2 of the buffer every time](https://github.com/microsoft/wil/blob/c4b5ea1fa3e2cd6a01aa8fdc3c5e509914e34a71/include/wil/win32_helpers.h#L526). But the other two APIs return the amount of required bytes if the buffer is too small, info that you discard, so your implementation is less efficient.

**Comment by rodrigocfd** [comment]
You're right, fixed.

**Comment by rodrigocfd** [comment]
If I got this right, I have to keep looping the function call until the char count is right... is that it?

**Comment by justanotheranonymoususer** [comment]
Yes. Say you run GetCurrentDirectory, it says - 260 chars is not enough, you need 300. Now you run it with 300. If it's fine, you're done. It will be fine if the path becomes smaller, too. But if it became bigger, you can get - 300 is not enough, you need 400. So now you need to try with 400, etc, until you succeeed.

**Comment by rodrigocfd** [comment]
The returned value of `GetCurrentDirectory` has three different meanings:

* if fails, zero;
* if buffer is not enough, the required size **plus** terminating null count;
* if succeeded, the number of written chars, **without** terminating null count.

I believe it's correct now.

**Comment by rodrigocfd** [comment]
Same for `ExpandEnvironmentStrings`.

**Comment by rodrigocfd** [comment]
`ExpandEnvironmentStrings` behaves differently: the returned value **always** counts the terminating null.

These inconsistencies are insane.

**Comment by justanotheranonymoususer** [comment]
Yes. What is "plus 1-char gap"?

**Comment by rodrigocfd** [comment]
> Yes. What is "plus 1-char gap"?

Unnecessary, in `ExpandEnvironmentStrings` we just need to check equality between returned length and current length.

**Comment by justanotheranonymoususer** [comment]
Use <= because it can be smaller too

**Comment by rodrigocfd** [comment]
> Use <= because it can be smaller too

Currect, the buffer may have shrunk.
